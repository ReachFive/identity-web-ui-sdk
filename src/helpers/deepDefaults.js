import union from 'lodash-es/union';
import isPlainObject from 'lodash-es/isPlainObject';
import isUndefined from 'lodash-es/isUndefined';

export function deepDefaults(object, ...defaults) {
    return defaults.length > 0
        ? deepDefaults(deepDefaultsR(object, defaults[0]), ...defaults.slice(1))
        : object;
}

function deepDefaultsR(object, defaults) {
    return union(
        Object.keys(object),
        Object.keys(defaults)
    ).reduce((acc, key) => {
        if (isPlainObject(object[key]) && isPlainObject(defaults[key])) {
            acc[key] = deepDefaultsR(object[key], defaults[key])
        } else if (isUndefined(object[key])) {
            acc[key] = defaults[key];
        } else {
            acc[key] = object[key];
        }
        return acc;
    }, {});
}
