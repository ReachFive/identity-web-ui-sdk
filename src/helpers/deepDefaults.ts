export function deepDefaults(object: Record<string, unknown>, ...defaults: Record<string, unknown>[]): Record<string, unknown> {
    return defaults.length > 0
        ? deepDefaults(deepDefaultsR(object, defaults[0]), ...defaults.slice(1))
        : object;
}

function deepDefaultsR(object: Record<string, unknown>, defaults: Record<string, unknown>): Record<string, unknown> {
    return [...new Set([
        ...Object.keys(object),
        ...Object.keys(defaults),
    ])]
        .reduce((acc, key) => {
            if (
                object[key] !== null && typeof object[key] === 'object'
                && defaults[key] !== null && typeof defaults[key] === 'object'
            ) {
                acc[key] = deepDefaultsR(object[key] as Record<string, unknown>, defaults[key] as Record<string, unknown>)
            } else if (object[key] === undefined) {
                acc[key] = defaults[key];
            } else {
                acc[key] = object[key];
            }
            return acc;
        }, {} as Record<string, unknown>);
}
