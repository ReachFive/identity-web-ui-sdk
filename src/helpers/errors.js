export class UserError extends Error {
    constructor(...params) {
        super(...params)
        this.isUserError = true;
    }
}
