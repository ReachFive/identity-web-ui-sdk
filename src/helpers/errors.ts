export interface AppError {
    errorId: string
    errorDescription: string
    error: string
    errorUserMsg?: string
    errorDetails?: string
    errorMessageKey?: string
}

export function isAppError(err: unknown): err is AppError {
    return typeof err === 'object' && err !== null && 'errorMessageKey' in err
}

export class UserError extends Error {
    isUserError = true

    constructor(message: string) {
        super(message)
        Object.setPrototypeOf(this, UserError.prototype);
    }

    static fromAppError(appError: AppError) {
        return new UserError(appError.errorUserMsg ?? appError.errorDescription ?? appError.error)
    }
}
