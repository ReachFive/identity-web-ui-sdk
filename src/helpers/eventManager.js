import pull from 'lodash-es/pull';
import { logError } from './logger';

export default class EventManager {
    constructor() {
        this.listeners = {};
    }

    fire(name, data) {
        getListeners(this.listeners, name).forEach(listener => {
            try {
                listener(data);
            } catch (e) {
                logError(e);
            }
        });
    }

    on(name, listener) {
        getListeners(this.listeners, name).push(listener);
    }

    off(name, listener) {
        pull(getListeners(this.listeners, name), listener);
    }
}

function getListeners(listeners, name) {
    return listeners[name] || (listeners[name] = []);
}
