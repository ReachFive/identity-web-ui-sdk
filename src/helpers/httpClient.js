import { toQueryString } from './queryString'
import { camelCaseProperties, snakeCaseProperties } from './transformObjectProperties';
import isEmpty from 'lodash-es/isEmpty';

export function createHttpClient(config) {
    function get(path, params) {
        return request(path, { ...params, method: 'GET' });
    }

    function post(path, params) {
        return request(path, { ...params, method: 'POST' });
    }

    function request(path, params) {
        const { method = 'GET', query = {}, body, accessToken = null, withCookies = false } = params;

        const fullPath = query && !isEmpty(query) ? `${path}?${toQueryString(query)}` : path;

        const url = fullPath.startsWith('http') ? fullPath : config.baseUrl + fullPath;

        const fetchOptions = {
            method,
            headers: {
                ...(accessToken && { Authorization: 'Bearer ' + accessToken }),
                ...(config.language && { 'Accept-Language': config.language }),
                ...(body && { 'Content-Type': 'application/json;charset=UTF-8' })
            },
            ...(withCookies && config.acceptCookies && { credentials: 'include' }),
            ...(body && { body: JSON.stringify(snakeCaseProperties(body)) })
        };

        return rawRequest(url, fetchOptions);
    }

    return { get, post, request };
}

export async function rawRequest(url, fetchOptions) {
    return fetch(url, fetchOptions).then(response => {
        if (response.status !== 204) {
            const dataP = response.json().then(camelCaseProperties);

            return response.ok ? dataP : dataP.then(data => Promise.reject(data));
        }
    });
}
