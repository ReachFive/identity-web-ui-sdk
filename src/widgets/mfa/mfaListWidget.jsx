import React from 'react';
import compose from '@hypnosphi/recompose/compose';

import { Info } from '../../components/miscComponent';
import { Card } from '../../components/form/cardComponent';
import { createWidget } from "../../components/widget/widget";
import {withConfig, withI18n, withTheme} from '../../components/widget/widgetContext'
import styled from "styled-components";

import { ReactComponent as Envelope } from '../../icons/envelope.svg'
import { ReactComponent as CommentAltDots } from '../../icons/comment-alt-dots.svg'

const iconStyle = props => `
    width: ${props.theme.get('fontSize') * 2}px;
    height: ${props.theme.get('fontSize') * 2}px;
    fill: ${props.theme.get('textColor')};
    flex-shrink: 0;
`

const EmailIcon = withTheme(styled(Envelope)`${props => iconStyle(props)}`);
const SmsIcon = withTheme(styled(CommentAltDots)`${props => iconStyle(props)}`);

const credentialIconByType = type => {
    switch (type) {
        case 'email':
            return <EmailIcon />
        case 'sms':
            return <SmsIcon />
    }
}

const Credential = withTheme(styled(Card)`
    display: flex;
    flex-direction: row;
    align-items: center;
    color: ${props => props.theme.get('textColor')};
    line-height: 1.5;
`);

const CardContent = withTheme(styled.div`
    margin-left: ${props => props.theme.get('_blockInnerHeight')}px;
    white-space: initial;
`);

const dateFormat = (dateString, locales) => (
    (new Date(dateString)).toLocaleDateString(locales, { timeZone: 'UTC' })
)

export const MfaList = compose(withI18n, withConfig)((({ credentials, i18n, config }) => (
    <div>
        {(credentials.length === 0) && (
            <Info>{i18n('mfaList.noCredentials')}</Info>
        )}
        {credentials.map(({type, createdAt, friendlyName, email, phoneNumber}, index) => (
            <Credential key={`credential-${index}`}>
                {credentialIconByType(type)}
                <CardContent>
                    <div style={{fontWeight: 'bold'}}>{friendlyName}</div>
                    <div>{email || phoneNumber}</div>
                    <div>
                        <span>{i18n('mfaList.createdAt')}&nbsp;: </span>
                        <time dateTime={createdAt}>{dateFormat(createdAt, config.language)}</time>
                    </div>
                </CardContent>
            </Credential>
        ))}
    </div>
)));

export default createWidget({
    component: MfaList,
    prepare: (options, { apiClient }) => {
        return apiClient.listMfaCredentials(options.accessToken).then(credentials => ({
            ...options,
            ...credentials,
        }))
    }
});
