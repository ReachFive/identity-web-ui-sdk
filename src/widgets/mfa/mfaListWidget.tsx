import { MFA } from '@reachfive/identity-core';
import React, { useEffect } from 'react';

import { createWidget } from '../../components/widget/widget';

import { LoaderCircle, Mail, MessageSquareMore, X } from 'lucide-react';
import {
    AlertDialog,
    AlertDialogAction,
    AlertDialogCancel,
    AlertDialogContent,
    AlertDialogDescription,
    AlertDialogFooter,
    AlertDialogHeader,
    AlertDialogTitle,
    AlertDialogTrigger,
} from '../../components/ui/alert-dialog';
import { Button } from '../../components/ui/button';
import { useConfig } from '../../contexts/config';
import { useI18n } from '../../contexts/i18n';
import { useReachfive } from '../../contexts/reachfive.tsx';
import { dateFormat } from '../../helpers/utils.ts';
import { OnError, OnSuccess } from '../../types';

const credentialIconByType = (type: MFA.CredentialsResponse['credentials'][number]['type']) => {
    switch (type) {
        case 'email':
            return <Mail className="bg-background w-icon h-icon stroke-textColor" />;
        case 'sms':
            return <MessageSquareMore className="bg-background w-icon h-icon stroke-textColor" />;
    }
};

export interface MfaListProps {
    /**
     * The authorization credential JSON Web Token (JWT) used to access the ReachFive API, less than five minutes old.
     */
    accessToken: string;
    /**
     * Callback function called when the request has succeeded.
     */
    onSuccess?: OnSuccess;
    /**
     * Callback function called when the request has failed.
     */
    onError?: OnError;
    /**
     * Indicates whether delete mfa credential button is displayed
     */
    showRemoveMfaCredential?: boolean;
}

interface DeleteButtonProps {
    credential: MFA.Credential;
    isOpen: boolean;
    setIsOpen: React.Dispatch<React.SetStateAction<boolean>>;
    onDeleteCallback: (credential: MFA.Credential) => void;
    deleteConfirmationTitle: string;
    setDeleteConfirmationTitle: React.Dispatch<React.SetStateAction<string>>;
}

const DeleteButton = ({
    credential,
    isOpen,
    setIsOpen,
    onDeleteCallback,
    deleteConfirmationTitle,
    setDeleteConfirmationTitle,
}: DeleteButtonProps) => {
    const i18n = useI18n();

    return (
        <AlertDialog open={isOpen} onOpenChange={setIsOpen}>
            <AlertDialogTrigger asChild>
                <Button
                    variant="destructive"
                    size="icon"
                    className="ml-1"
                    onClick={_ =>
                        setDeleteConfirmationTitle(
                            credential.type === 'email'
                                ? 'mfa.remove.email'
                                : 'mfa.remove.phoneNumber'
                        )
                    }
                >
                    <X />
                </Button>
            </AlertDialogTrigger>
            <AlertDialogContent>
                <AlertDialogHeader>
                    <AlertDialogTitle>{i18n(deleteConfirmationTitle)}</AlertDialogTitle>
                    <AlertDialogDescription />
                </AlertDialogHeader>
                <AlertDialogFooter>
                    <AlertDialogCancel>{i18n('confirmation.cancel')}</AlertDialogCancel>
                    <AlertDialogAction
                        variant="destructive"
                        onClick={_ => onDeleteCallback(credential)}
                    >
                        {i18n('confirmation.yes')}
                    </AlertDialogAction>
                </AlertDialogFooter>
            </AlertDialogContent>
        </AlertDialog>
    );
};

export const MfaList = ({
    accessToken,
    showRemoveMfaCredential,
    onError = (() => {}) as OnError,
    onSuccess = (() => {}) as OnSuccess,
}: MfaListProps) => {
    const [isOpen, setIsOpen] = React.useState(false);
    const [loading, setLoading] = React.useState(true);
    const [deleteConfirmationTitle, setDeleteConfirmationTitle] = React.useState('');
    const [credentials, setCredentials] = React.useState<MFA.Credential[]>([]);
    const i18n = useI18n();
    const config = useConfig();
    const client = useReachfive();

    const fetchMfaCredentials = () => {
        setLoading(true);
        client
            .listMfaCredentials(accessToken)
            .then(mfaCredentialsResponse => {
                setCredentials(mfaCredentialsResponse.credentials);
                onSuccess({
                    name: 'mfa_credentials_listed',
                    credentials: mfaCredentialsResponse.credentials,
                });
            })
            .catch(onError)
            .finally(() => setLoading(false));
    };

    useEffect(() => {
        fetchMfaCredentials();
    }, [accessToken]);

    const onDelete = (credential: MFA.Credential) => {
        switch (credential.type) {
            case 'sms':
                client
                    .removeMfaPhoneNumber({
                        accessToken,
                        phoneNumber: credential.phoneNumber,
                    })
                    .then(() => {
                        fetchMfaCredentials();
                        onSuccess({ name: 'mfa_email_removed' });
                    })
                    .catch(onError);
                break;
            case 'email':
                client
                    .removeMfaEmail({ accessToken })
                    .then(() => {
                        fetchMfaCredentials();
                        onSuccess({ name: 'mfa_email_removed' });
                    })
                    .catch(onError);
                break;
        }
    };

    if (loading) {
        return <LoaderCircle className="animate-spin" />;
    }

    return (
        <div>
            {credentials.length === 0 && (
                <div className="mb-1 text-center text-textColor">
                    {i18n('mfaList.noCredentials')}
                </div>
            )}
            {credentials.map((credential, _) => (
                <div
                    id={`credential-${credential.friendlyName}`}
                    data-testid="credential"
                    key={`credential-${credential.friendlyName}`}
                    className={`flex flex-col ${isOpen ? 'opacity-15' : ''}`}
                >
                    <div className="flex flex-row items-center rounded">
                        <div className="box-border flex flex-row items-center align-middle border border-solid rounded basis-full whitespace-nowrap p-generic">
                            {credentialIconByType(credential.type)}
                            <div className="ml-innerBlock w-max justify-items-stretch text-textColor">
                                <div className="font-bold">{credential.friendlyName}</div>
                                <div>
                                    {MFA.isEmailCredential(credential)
                                        ? credential.email
                                        : MFA.isPhoneCredential(credential)
                                          ? credential.phoneNumber
                                          : 'N/A'}
                                </div>
                                <div>
                                    <span>{i18n('mfaList.createdAt')}&nbsp;: </span>
                                    <time dateTime={credential.createdAt}>
                                        {dateFormat(credential.createdAt, config.language)}
                                    </time>
                                </div>
                            </div>
                        </div>
                        {showRemoveMfaCredential && (
                            <DeleteButton
                                isOpen={isOpen}
                                setIsOpen={setIsOpen}
                                onDeleteCallback={onDelete}
                                deleteConfirmationTitle={deleteConfirmationTitle}
                                setDeleteConfirmationTitle={setDeleteConfirmationTitle}
                                credential={credential}
                            />
                        )}
                    </div>
                </div>
            ))}
        </div>
    );
};

export type MfaListWidgetProps = {
    /**
     * The authorization credential JSON Web Token (JWT) used to access the ReachFive API, less than five minutes old.
     */
    accessToken: string;
    /**
     * Callback function called when the request has succeeded.
     */
    onSuccess?: OnSuccess;
    /**
     * Callback function called when the request has failed.
     */
    onError?: OnError;
    /**
     * Indicates whether delete mfa credential button is displayed
     */
    showRemoveMfaCredential?: boolean;
};

export default createWidget<MfaListWidgetProps, MfaListProps>({
    component: MfaList,
});
