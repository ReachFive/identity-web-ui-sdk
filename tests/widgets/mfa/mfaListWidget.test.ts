/**
 * @jest-environment jsdom
 */

import { beforeEach, describe, expect, jest, test } from '@jest/globals';
import '@testing-library/jest-dom/jest-globals';
import { render, screen, waitFor } from '@testing-library/react';
import 'jest-styled-components';

import type { Client, MFA } from '@reachfive/identity-core';

import type { I18nMessages } from '../../../src/core/i18n';
import type { Config, OnError, OnSuccess } from '../../../src/types';

import { AppError } from '../../../src/helpers/errors';
import mfaListWidget from '../../../src/widgets/mfa/mfaListWidget';

const defaultConfig: Config = {
    clientId: 'local',
    domain: 'local.reach5.net',
    sso: false,
    sms: false,
    webAuthn: false,
    language: 'fr',
    pkceEnforced: false,
    isPublic: true,
    socialProviders: ['facebook', 'google'],
    customFields: [],
    resourceBaseUrl: 'http://localhost',
    mfaSmsEnabled: true,
    mfaEmailEnabled: true,
    rbaEnabled: false,
    consentsVersions: {},
    passwordPolicy: {
        minLength: 8,
        minStrength: 2,
        allowUpdateWithAccessTokenOnly: true,
    },
};

const defaultI18n: I18nMessages = {};

describe('Snapshot', () => {
    const generateSnapshot =
        (
            options: Partial<Parameters<typeof mfaListWidget>[0]>,
            config: Partial<Config> = {},
            credentials: (MFA.EmailCredential | MFA.PhoneCredential)[]
        ) =>
        async () => {
            // @ts-expect-error partial Client
            const apiClient: Client = {
                listMfaCredentials: jest
                    .fn<Client['listMfaCredentials']>()
                    .mockResolvedValue({ credentials }),
            };

            const widget = await mfaListWidget(
                { accessToken: 'azerty', ...options },
                { apiClient, config: { ...defaultConfig, ...config }, defaultI18n }
            );

            await waitFor(async () => {
                const { container, rerender } = await render(widget);

                await waitFor(() => expect(apiClient.listMfaCredentials).toHaveBeenCalled());

                rerender(widget);

                expect(container).toMatchSnapshot();
            });
        };

    test('empty', generateSnapshot({}, undefined, []));

    test(
        'basic',
        generateSnapshot({}, undefined, [
            {
                type: 'sms',
                phoneNumber: '33612345678',
                friendlyName: '33612345678',
                createdAt: '2022-09-21',
            },
            {
                type: 'email',
                email: 'root@reach5.co',
                friendlyName: 'identifier',
                createdAt: '2022-09-21',
            },
        ])
    );
});

describe('DOM testing', () => {
    const listMfaCredentials = jest.fn<Client['listMfaCredentials']>();

    const onError = jest.fn<OnError>();
    const onSuccess = jest.fn<OnSuccess>();

    beforeEach(() => {
        onError.mockClear();
        onSuccess.mockClear();
        listMfaCredentials.mockClear();
    });

    const generateComponent = async (
        options: Partial<Parameters<typeof mfaListWidget>[0]>,
        config: Partial<Config> = {}
    ) => {
        // @ts-expect-error partial Client
        const apiClient: Client = {
            listMfaCredentials,
        };
        const result = await mfaListWidget(
            { accessToken: 'azerty', onError, onSuccess, ...options },
            { apiClient, config: { ...defaultConfig, ...config }, defaultI18n }
        );
        return await waitFor(async () => {
            return render(result);
        });
    };

    describe('mfaCredentials', () => {
        test('no credentials', async () => {
            listMfaCredentials.mockResolvedValue({ credentials: [] });
            await generateComponent({}, defaultConfig);

            const noCredentials = screen.queryByText('mfaList.noCredentials');
            expect(noCredentials).toBeInTheDocument();

            const credential = screen.queryAllByTestId('credential');
            expect(credential).toHaveLength(0);

            expect(onSuccess).toBeCalledWith(
                expect.objectContaining({
                    credentials: expect.arrayContaining([]),
                    name: 'mfa_credentials_listed',
                })
            );
            expect(onError).not.toBeCalled();
        });

        test('with credentials', async () => {
            listMfaCredentials.mockResolvedValue({
                credentials: [
                    {
                        type: 'sms',
                        phoneNumber: '33612345678',
                        friendlyName: '33612345678',
                        createdAt: '2022-09-21',
                    } as MFA.PhoneCredential,
                    {
                        type: 'email',
                        email: 'root@reach5.co',
                        friendlyName: 'identifier',
                        createdAt: '2022-09-21',
                    } as MFA.EmailCredential,
                ],
            });
            await generateComponent({}, defaultConfig);

            const noCredentials = screen.queryByText('mfaList.noCredentials');
            expect(noCredentials).not.toBeInTheDocument();

            const credential = screen.queryAllByTestId('credential');
            expect(credential).toHaveLength(2);

            expect(onSuccess).toBeCalledWith(
                expect.objectContaining({
                    credentials: expect.arrayContaining([
                        {
                            type: 'sms',
                            phoneNumber: '33612345678',
                            friendlyName: '33612345678',
                            createdAt: '2022-09-21',
                        },
                        {
                            type: 'email',
                            email: 'root@reach5.co',
                            friendlyName: 'identifier',
                            createdAt: '2022-09-21',
                        },
                    ]),
                    name: 'mfa_credentials_listed',
                })
            );
            expect(onError).not.toBeCalled();
        });

        test('api error', async () => {
            const error: AppError = {
                errorId: '0',
                error: 'unexpected_error',
                errorDescription: 'Unexpected error',
            };
            listMfaCredentials.mockRejectedValue(error);
            await expect(generateComponent({}, defaultConfig)).rejects.toEqual(error);
            expect(onSuccess).not.toBeCalled();
            expect(onError).toBeCalledWith(error);
        });
    });
});
