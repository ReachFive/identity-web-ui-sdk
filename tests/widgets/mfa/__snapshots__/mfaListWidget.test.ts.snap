// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot basic 1`] = `
.c0 {
  font-size: 14px;
  -webkit-transition: -webkit-transform 400ms ease, opacity 400ms ease;
  -webkit-transition: transform 400ms ease, opacity 400ms ease;
  transition: transform 400ms ease, opacity 400ms ease;
  opacity: 0;
  padding: 20px;
  border-radius: 3px;
  background-color: #ffffff;
  max-width: 400px;
  box-sizing: border-box;
  margin: 0 auto;
}

<div>
  <div
    class="c0"
  >
    <div>
      <div
        class="flex flex-col "
        data-testid="credential"
        id="credential-0"
      >
        <div
          class="flex flex-row items-center rounded"
        >
          <div
            class="flex flex-row basis-full items-center rounded align-middle whitespace-nowrap box-border p-generic border-solid border-generic border-border"
          >
            <svg
              class="lucide lucide-message-square-more bg-background w-icon h-icon stroke-textColor"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M21 15a2 2 0 0 1-2 2H7l-4 4V5a2 2 0 0 1 2-2h14a2 2 0 0 1 2 2z"
              />
              <path
                d="M8 10h.01"
              />
              <path
                d="M12 10h.01"
              />
              <path
                d="M16 10h.01"
              />
            </svg>
            <div
              class="ml-innerBlock w-max justify-items-stretch text-textColor"
            >
              <div
                class="font-bold"
              >
                identifier
              </div>
              <div>
                33612345678
              </div>
              <div>
                <span>
                  mfaList.createdAt
                   : 
                </span>
                <time
                  datetime="2022-09-21"
                >
                  21/09/2022
                </time>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="flex flex-col "
        data-testid="credential"
        id="credential-1"
      >
        <div
          class="flex flex-row items-center rounded"
        >
          <div
            class="flex flex-row basis-full items-center rounded align-middle whitespace-nowrap box-border p-generic border-solid border-generic border-border"
          >
            <svg
              class="lucide lucide-mail bg-background w-icon h-icon stroke-textColor"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="16"
                rx="2"
                width="20"
                x="2"
                y="4"
              />
              <path
                d="m22 7-8.97 5.7a1.94 1.94 0 0 1-2.06 0L2 7"
              />
            </svg>
            <div
              class="ml-innerBlock w-max justify-items-stretch text-textColor"
            >
              <div
                class="font-bold"
              >
                identifier
              </div>
              <div>
                root@reach5.co
              </div>
              <div>
                <span>
                  mfaList.createdAt
                   : 
                </span>
                <time
                  datetime="2022-09-21"
                >
                  21/09/2022
                </time>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Snapshot empty 1`] = `
.c0 {
  font-size: 14px;
  -webkit-transition: -webkit-transform 400ms ease, opacity 400ms ease;
  -webkit-transition: transform 400ms ease, opacity 400ms ease;
  transition: transform 400ms ease, opacity 400ms ease;
  opacity: 0;
  padding: 20px;
  border-radius: 3px;
  background-color: #ffffff;
  max-width: 400px;
  box-sizing: border-box;
  margin: 0 auto;
}

<div>
  <div
    class="c0"
  >
    <div>
      <div
        class="text-textColor mb-1 text-center"
      >
        mfaList.noCredentials
      </div>
    </div>
  </div>
</div>
`;
